src/fallbacks.jl

```diff
@@ -59,20 +59,18 @@ function rows(x::T) where {T}
         cols = columns(x)
         return RowIterator(cols, rowcount(cols))
     elseif IteratorInterfaceExtensions.isiterable(x)
-        return IteratorWrapper(IteratorInterfaceExtensions.getiterator(x))
+        return nondatavaluerows(x)
     end
     throw(ArgumentError("no default `Tables.rows` implementation for type: $T"))
 end
 
 # build columns from rows
-haslength(L) = L isa Union{Base.HasShape, Base.HasLength}
-
 """
     Tables.allocatecolumn(::Type{T}, len) => returns a column type (usually AbstractVector) w/ size to hold `len` elements
 
     Custom column types can override with an appropriate "scalar" element type that should dispatch to their column allocator.
 """
-allocatecolumn(T, len) = Vector{T}(undef, len)
+allocatecolumn(T, len) = DataAPI.defaultarray(T, 1)(undef, len)
```

Comment:
Should `defaultarray` replace `allocatecolumn` at some point?

Reply:
they're slightly different; `defaultarray` just returns the array type for a given scalar type; allocatecolumn is for actually creating a column. We can probably just replace uses of `allocatecolumn` with `defaultarray(T)(undef, len)`

Reply:
Yes, that's what I had in mind.